src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.ParsedSyntaxTree.cs

```diff
@@ -94,6 +110,8 @@ public override CSharpParseOptions Options
                 }
             }
 
+            public override ImmutableDictionary<string, ReportDiagnostic> DiagnosticOptions => _diagnosticOptions;
```

Comment:
could these just live on the CSharpParseOptions  object?  Then there would be no need to for any of these changes just to flow these options along with the options we already know how to create and flow along.

Reply:
I don't think that's a good idea. Changing the parse options should indicate that a new parse is needed, but changing the diagnostic options should not. In theory, that would cause every single change to an editorconfig file to reparse every file from scratch.

Reply:
> but changing the diagnostic options should not. 

1. what will happen if you change the diagnotsic options.  certianly you'll need new diagnostics.  Do you intend to always compute all the diagnostics, but just use this mapping to determine which to filter down to?

2. I don't necessarily agree that changing parse options means you must parse.  Changing parse options means you shoudl reparse if you got new options that could actually affect parsing.  If hte data in the ParseOptions that affects parsing is changed, then you reparse.

nearly 100% of the time parse options will not change, so this will just be a reference equals that shortcircuits.  IN the time when parse options actually have changed, we can just check and see what is different and only parse based on what is different.

Reply:
From my reading there's like a half dozen fields in this guy to check.  i would still far prefer that approach :)   Thoughts?

Reply:
i feel like that's an internal implementation detail.  the contract should simply be:  you get the syntax tree you should get based on these options.  It's regrettable we called these ParseOptions instead of SyntaxTreeOptions, which would fit better (especially given that we have CompilationOptions).  The compiler can choose whatever strategy it wants with new options, just as it it can do things like optimize how it builds the decl table depending on what actually changed, rather than building the full one all the time.  I'm curious what the thoughts are here from others who helped design the original compilation model APIs.  

To me, i always thought of ParseOptions as simply: the options that follow a document, whereas CompilationOptions are the options that follow a Project.  So it makes total sense to me that we would continue just using those types.  

@gafter @mattwar @jasonmalinowski  do you have any thoughts/opinions here?